abi.html

```diff
@@ -662,40 +693,110 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
-     represented as a <code>ptrdiff_t</code>.
-</dl>
+Note that by <code>[dcl.init]</code>, "zero initialization" of a data
+member pointer object stores a null pointer value into it.  Under this
+representation, that value has a non-zero bit representation.  On most
+modern platforms, data member pointers are the only type with this
+property.
+
+<p>
+Base-to-derived and derived-to-base conversions of a non-null data member
+pointer can be performed by adding or subtracting (respectively) the static
+offset of the base within the derived class.  The C++ standard does not
+permit base-to-derived and derived-to-base conversions of member pointers
+to cross a <code>virtual</code> base relationship, and so a static offset
+is always known.
+
+<a name="member-function-pointers"></a>
+<h4><a href="#member-function-pointers"> 2.3.2 Member Function Pointers </a></h4>
+
+<p>
+The basic ABI properties of member function pointer types are those of
+the following class:
+
+<pre>
+  struct {
+    fnptr_t ptr;
+    ptrdiff_t adj;
+  };
+</pre>
+
+where <code>fnptr_t</code> is the appropriate function-pointer type
+for the member type.
+
+<p>
+A member function pointer for a non-virtual function is represented
+with <code>ptr</code> set to a function pointer to the function,
+using the base ABI's representation of function pointers.
 
 <p>
-It has the size, data size, and alignment
-of a class containing those two members, in that order.
-(For 64-bit Itanium, that will be 16, 16, and 8 bytes respectively.)
+A member function pointer for a virtual function is represented
+with <code>ptr</code> set to 1 plus the virtual table offset of
+the function (in bytes), converted to a function pointer as if by <code>reinterpret_cast&lt;fnptr_t&gt;(uintfnptr_t(1 + offset))</code>,
+where <code>uintfnptr_t</code> is an unsigned integer of the same
+size as <code>fnptr_t</code>.
 
+<p>
+In both cases, <code>adj</code> stores the offset (in bytes) which
+must be added to a pointer to the base type before the call can be
+made.  For a virtual member function pointer, the v-table is loaded
+from the adjusted address.
+
+<p>
+The representation of virtual member function pointers requires the
+representation of a function pointer to a non-static member function
+to never have its lowest bit set.  On most platforms, this is either
+always true, or it can be made true at little cost.  (For example,
+on most platforms a function pointer is just the address of the
+first instruction in the function.  Many architectures align all
+instructions to at least 2 bytes; even on architectures where that's
+not true, or where functions aren't byte-addressed, the compiler can
+just increase the alignment of non-static member functions until the
+low bit of the address is always reliably zero.)  However, some platforms
+use the low bit of a function pointer for special purposes, such as to
+distinguish THUMB functions on ARM.  Such platforms must use a slightly
```

Comment:
"must" seems a bit strong. How about "For such platforms, we recommend using a slightly different representation" or similar?

Reply:
I ended up rewriting this whole section.